### ISSUE TITLE
Check that type hints match when implementing an ABC

### ISSUE THREAD

**Issue by MartinBernstorff** [microsoft/pyright#5382]
Specifically, this would allow for interface-like programming in Python.

Let's start with the challenge:

```
class Interface(ABC):
    @abstractmethod
    def test_method(self) -> int:
        ...


class ImplementsInterface(Interface):
    def test_method(self) -> None:
        return None
```

In this situation, pyright does not raise an error, but mypy does!

```bash
❯ mypy abc_test2.py 
abc_test2.py:12: error: Return type "None" of "test_method" incompatible with return type "int" in supertype "Interface"  [override]
Found 1 error in 1 file (checked 1 source file)

❯ pyright abc_test2.py 
WARNING: there is a new pyright version available (v1.1.305 -> v1.1.315).
Please install the new version or set PYRIGHT_PYTHON_FORCE_VERSION to `latest`

No configuration file found.
pyproject.toml file found at /Users/au484925/Desktop/psycop-common.
Loading pyproject.toml file at /Users/au484925/Desktop/psycop-common/pyproject.toml
Assuming Python version 3.10
Searching for source files
Found 1 source file
pyright 1.1.305
0 errors, 0 warnings, 0 informations 
Completed in 0.467sec
```

I'm not experienced enough to know what the "correct" behaviour is here. On the one hand, I guess you're "overloading" the abstractmethod as you're implementing it, and in that case, it's allowed to change its return type. 

On the other, I think it's terrible practice, and would for it to be flagged as an error. 

Would love to hear what you think! 

And thanks for extreme responsiveness and an awesome tool 👍

**Comment by erictraut** [comment]
Pyright will detect and report incompatible method overrides if you enable the `reportIncompatibleMethodOverride` check. This is off by default in "basic" type-checking mode but is enabled by default in "strict" mode. For a full list of diagnostic checks, refer to [this documentation](https://microsoft.github.io/pyright/#/configuration?id=type-check-diagnostics-settings).
